CMakeLists.txt

```diff
@@ -1,72 +1,101 @@
+add_library(ensmallen INTERFACE)
+target_include_directories(ensmallen INTERFACE
+  $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+  $<INSTALL_INTERFACE:include>)
+
+# Set warning flags for target.
+if(MSVC)
+  target_compile_options(ensmallen INTERFACE /Wall )
 else()
-  # set required standard to c++11
-  set(CMAKE_CXX_STANDARD 11)
-  set(CMAKE_CXX_STANDARD_REQUIRED ON)
-endif ()
-
-# Detect OpenMP support in a compiler. If the compiler supports OpenMP, flags
-# to compile with OpenMP are returned and added for compilation.
-if (USE_OPENMP)
-  find_package(OpenMP)
-endif ()
-
-if (OPENMP_FOUND)
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
-else ()
-  # Disable warnings for all the unknown OpenMP pragmas.
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas")
-endif ()
-
-# Set the CFLAGS and CXXFLAGS depending on the options the user specified.
-if(CMAKE_COMPILER_IS_GNUCC OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wpedantic -Wunused-parameter")
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wpedantic -Wunused-parameter")
+  target_compile_options(ensmallen INTERFACE -Wall -Wpedantic -Wunused-parameter)
 endif()
 
-# The only dependency we need is Armadillo.
-#
-# We keep the minimum version in sync with mlpack, otherwise we could have
-# irritating compatibility issues.
+# Find OpenMP and link it.
+if(USE_OPENMP)
+  if(NOT TARGET OpenMP::OpenMP_CXX)
+    find_package(Threads REQUIRED)
+    add_library(OpenMP::OpenMP_CXX IMPORTED INTERFACE)
+    set_property(TARGET OpenMP::OpenMP_CXX
+        PROPERTY INTERFACE_COMPILE_OPTIONS ${OpenMP_CXX_FLAGS})
+    # Only works if the same flag is passed to the linker; use CMake 3.9+ otherwise (Intel, AppleClang).
+    set_property(TARGET OpenMP::OpenMP_CXX
+        PROPERTY INTERFACE_LINK_LIBRARIES ${OpenMP_CXX_FLAGS} Threads::Threads)
+  endif()
+  target_link_libraries(ensmallen INTERFACE OpenMP::OpenMP_CXX)
+endif()
+
+# Find Armadillo and link it.
 find_package(Armadillo 8.400.0 REQUIRED)
-include_directories(BEFORE "${ARMADILLO_INCLUDE_DIR}")
-include_directories(BEFORE "${CMAKE_SOURCE_DIR}/include/")
+target_link_libraries(ensmallen INTERFACE Armadillo::Armadillo)
+
+# Set helper variables for creating the version, config and target files.
+include(CMakePackageConfigHelpers)
+set(ENSMALLEN_CMAKE_DIR "lib/cmake/ensmallen" CACHE STRING
+      "Installation directory for cmake files, relative to ${CMAKE_INSTALL_PREFIX}.")
+set(version_config "${PROJECT_BINARY_DIR}/ensmallen-config-version.cmake")
+set(project_config "${PROJECT_BINARY_DIR}/ensmallen-config.cmake")
+set(targets_export_name ensmallen-targets)
 
-# Install the headers to the correct location.
+# Generate the version, config and target files into the build directory.
+write_basic_package_version_file(${version_config}
+    VERSION ${VERSION}
+    COMPATIBILITY AnyNewerVersion)
```

Comment:
Should we use `ARCH_INDEPENDENT` here?  https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#generating-a-package-version-file

Reply:
This isn't an available option for CMake 3.3.2, I believe it has been added in one of the latest versions

Reply:
Ah, got it, thanks for the clarification.  Looks like `ARCH_INDEPENDENT` was added in 3.14.  Do you think we should do something like this?

```
if (${CMAKE_VERSION} VERSION_LESS "3.14.0")
  write_basic_package_version_file( ... )
else ()
  write_basic_package_version_file( ... ARCH_INDEPENDENT ... )
endif ()
```

I'm not sure how important the `ARCH_INDEPENDENT` really is.